lang/cs/Org.Apache.REEF.IMRU.Tests/TestContextManager.cs

```diff
@@ -0,0 +1,337 @@
+                var c = new MyActiveContext(ContextIdPrefix + i);
+                contextManager.Add(c);
+            }
+            Assert.True(contextManager.AllContextsReceived);
+            Assert.Equal(5, contextManager.NumberOfActiveContext);
+            var e = new MyFailedEvaluator(ContextIdPrefix + "3");
+            contextManager.RemovedFailedContextInFailedEvaluator(e);
+            Assert.Equal(1, contextManager.NumberOfMissingContexts);
+            contextManager.Remove(ContextIdPrefix + "4");
+            Assert.Equal(3, contextManager.NumberOfActiveContext);
+        }
+
+        /// <summary>
+        /// Test remove a failed evaluator which has two contexts associated.
+        /// In current IMRU driver, assume there is only one context associated to the IFailedEvalutor
+        /// </summary>
+        [Fact]
+        public void TestRemoveFailedEvaluatorWithTwoContexts()
+        {
+            var contextManager = new ContextManager(5);
+            for (int i = 0; i < 5; i++)
+            {
+                var c = new MyActiveContext(ContextIdPrefix + i);
+                contextManager.Add(c);
+            }
+
+            var e = new MyFailedEvaluatorWithTwoContexts();
+            Action remove = () => contextManager.RemovedFailedContextInFailedEvaluator(e);
+            Assert.Throws<IMRUSystemException>(remove);
+        }
+
+        /// <summary>
+        /// Test remove a failed evaluator which has a context but it doesn't exist.
+        /// </summary>
+        [Fact]
+        public void TestRemoveFailedEvaluatorWithNoExistsContexts()
+        {
+            var contextManager = new ContextManager(5);
+            for (int i = 0; i < 5; i++)
+            {
+                var c = new MyActiveContext(ContextIdPrefix + i);
+                contextManager.Add(c);
+            }
+
+            var e = new MyFailedEvaluator(ContextIdPrefix + "5");
+            Action remove = () => contextManager.RemovedFailedContextInFailedEvaluator(e);
+            Assert.Throws<SystemException>(remove);
+        }
+
+        /// <summary>
+        /// Test remove a failed evaluator which has no context associated.
+        /// The scenario may happen when an evaluator failed but context has not created yet. 
+        /// </summary>
+        [Fact]
+        public void TestRemoveFailedEvaluatorWithNoContext()
+        {
+            var contextManager = new ContextManager(5);
+            for (int i = 0; i < 5; i++)
+            {
+                var c = new MyActiveContext(ContextIdPrefix + i);
+                contextManager.Add(c);
+            }
+
+            var e = new MyFailedEvaluatorWithNoContext();
+            contextManager.RemovedFailedContextInFailedEvaluator(e);
+            Assert.Equal(5, contextManager.NumberOfActiveContext);
+        }
+
+        /// <summary>
+        /// An implementation of IFailedEvaluator for testing
+        /// </summary>
+        private class MyFailedEvaluator : IFailedEvaluator
```

Comment:
Do we need to keep three different implementations of `IFailedEvaluator` here? They differ only in the number of contexts `FailedContexts` returns, so maybe we could replace these classes with one `MyFailedEvaluatorWithNContexts` and pass number of contexts as second constructor parameter?